Automattic/newspack-blocks
# Block caching

### All Submissions:






### Changes proposed in this Pull Request:



This PR introduces caching for the Posts and Carousel blocks. This is off by default for now, so I can run this on a dozen or so sites for a while and make sure it's solid, and we'll switch it to on by default at some point before elections.

### How to test the changes in this Pull Request:

You'll probably want to use a WP Cloud site for testing so that you have a good object cache running, but it may work OK on your local site. It definitely shouldn't cause issues or break anything if an object cache isn't present.

1. To your wp-config add: `define( 'NEWSPACK_LOG_LEVEL', 2 );` and `define( 'NEWSPACK_BLOCKS_CACHE_BLOCKS', true );`.
2. Visit a Page with posts blocks and/or carousel blocks. On the first visit you should see something like this for each block on the page in the logs:
```
[14-Jun-2024 19:43:33 UTC] [74084][NEWSPACK][Newspack_Blocks_Caching::debug_log]: Checking cache for: np_cached_block_f87a58c1699b8f382352a28962df71a8
[14-Jun-2024 19:43:33 UTC] [74084][NEWSPACK][Newspack_Blocks_Caching::debug_log]: Cached data not found for: np_cached_block_f87a58c1699b8f382352a28962df71a8
[14-Jun-2024 19:43:33 UTC] [74084][NEWSPACK][Newspack_Blocks_Caching::debug_log]: Caching block: np_cached_block_f87a58c1699b8f382352a28962df71a8
```
3. Refresh the page. It should load much faster but should otherwise look exactly the same. In the logs you should see something like:
```
[14-Jun-2024 19:43:33 UTC] [74084][NEWSPACK][Newspack_Blocks_Caching::debug_log]: Checking cache for: np_cached_block_ae84d66a8217d97cc4e70226086bb965
[14-Jun-2024 19:43:33 UTC] [74084][NEWSPACK][Newspack_Blocks_Caching::debug_log]: Serving cached block: np_cached_block_ae84d66a8217d97cc4e70226086bb965
```
4. Wait 5 minutes. This should invalidate the cache and force a re-cache of everything.
5. Add a Posts block to a widget area (e.g. sidebar). Visit a post, observe the block gets cached. Visit a different post, observe the cached version of the block gets served. This does not happen on Pages but should happen on Posts. See the explanation of the heuristic in the code.
6. `wp cache flush` to flush the cache if you want to start from the top.

### Other information:
